internals/daemon/api_exec_test.go

```diff
@@ -441,6 +443,53 @@ func (s *execSuite) TestUserGroupError(c *C) {
 	c.Check(execResp.Result["message"], Matches, ".*must specify user, not just group.*")
 }
 
+// TestExecChangeReady simulates the scenario where the change is ready before the websocket
+// connection is established, so the connection should fail.
+func (s *execSuite) TestExecChangeReady(c *C) {
+	httpResp, execResp := execRequest(c, &client.ExecOptions{
+		Command: []string{"echo", "foo"},
+	})
+	c.Assert(httpResp.StatusCode, Equals, http.StatusAccepted)
+	body := new(bytes.Buffer)
+	err := json.NewEncoder(body).Encode(execResp)
+	c.Assert(err, IsNil)
+
+	changeID := execResp.Change
+	c.Assert(changeID, Not(Equals), "")
+
+	st := s.daemon.overlord.State()
+	st.Lock()
+	change := st.Change(changeID)
+	c.Assert(change, NotNil)
+	c.Assert(len(change.Tasks()), Equals, 1)
+	// Set the change as failed and set the error on the task.
+	change.SetStatus(state.ErrorStatus)
+	change.Tasks()[0].Errorf("something went wrong")
+	change.Tasks()[0].SetStatus(state.ErrorStatus)
+	st.Unlock()
+
+	taskID, ok := execResp.Result["task-id"].(string)
+	c.Assert(ok, Equals, true)
+
+	vars := map[string]string{"task-id": taskID, "websocket-id": "control"}
+	restoreMuxVars := FakeMuxVars(func(*http.Request) map[string]string {
+		return vars
+	})
+	defer restoreMuxVars()
+
+	websocketCmd := apiCmd("/v1/tasks/{task-id}/websocket/{websocket-id}")
+	req, err := http.NewRequest("GET", fmt.Sprintf("/v1/tasks/%s/websocket/%s", taskID, "control"), nil)
+	c.Assert(err, IsNil)
+	rsp := v1GetTaskWebsocket(websocketCmd, req, nil).(websocketResponse)
+	rec := httptest.NewRecorder()
+	rsp.ServeHTTP(rec, req)
+
+	c.Check(rec.Code, Equals, 500)
+	err = rsp.connect(req, rec, rsp.task, rsp.websocketID)
+	c.Assert(err, NotNil)
+	c.Check(err.Error(), Matches, `.*cannot perform the following tasks:\n.*something went wrong.*`)
```

Comment:
rsp.ServeHTTP has already called connect internally. So we should be checking for rec.Result() body to be the error instead.